### PR TITLE
OWL implementation of the calendar view

### PR DESCRIPTION
**Pull request by mcm-odoo**
Enterprise: odoo/enterprise#31291

### CODE REVIEW

File: addons/web/static/src/views/calendar/calendar_model.js [hunk 1]
```diff
@@ -472,10 +474,18 @@ export class CalendarModel extends Model {
...
-        const showTime =
+        let showTime =
             !(fieldMapping.all_day && rawRecord[fieldMapping.all_day]) &&
             scale === "month" &&
             startType !== "date";
+
+        // An "all day" event without the "all_day" option is not considered
+        // as a 24h day. It's just a part of the day (by default: 7h-19h).
+        // So it should be shown as an allday event (with the time hidden)
+        if (scale === "month" && !this.hasAllDaySlot && start.hour === 7 && end.hour === 19) {
```
Comment (brboi) [inline addons/web/static/src/views/calendar/calendar_model.js:510]: what about minutes/seconds/milliseconds :-D

File: addons/web/static/src/views/calendar/calendar_model.js [hunk 2]
```diff
@@ -486,6 +486,11 @@ export class CalendarModel extends Model {
             showTime = false;
         }
 
+        //should not display a time for multiple days event
+        if (scale === "month" && (end - start) / (1000 * 3600 * 24) > 1) {
```
Comment (brboi) [inline addons/web/static/src/views/calendar/calendar_model.js:515]: only check if day number is different

File: addons/web/static/src/views/calendar/calendar_arch_parser.js [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+                    break;
+                }
+            }
+        });
+
+        return {
+            canCreate,
+            canDelete,
+            canEdit,
```
Comment (aab-odoo) [inline addons/web/static/src/views/calendar/calendar_arch_parser.js:182]: in other view, I think we regroup those in `activeActions`

File: addons/web/static/src/views/calendar/calendar_common/calendar_common_popover.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
+/** @odoo-module **/
+
+import { Dialog } from "@web/core/dialog/dialog";
+import { is24HourFormat } from "@web/core/l10n/dates";
+import { Field } from "../../fields/field";
+import { Record } from "../../record";
```
Comment (aab-odoo) [inline addons/web/static/src/views/calendar/calendar_common/calendar_common_popover.js:6]: I would use absolute paths for everything except calendar stuff

File: addons/web/static/src/views/calendar/calendar_common/calendar_common_renderer.js [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+    }
+    onEventDrop(info) {
+        this.fc.api.unselect();
+        this.props.model.updateRecord(this.FCEventToRecord(info.event), { moved: true });
+    }
+    onEventResize(info) {
+        this.fc.api.unselect();
+        this.props.model.updateRecord(this.FCEventToRecord(info.event));
+    }
+    FCEventToRecord(event) {
```
Comment (aab-odoo) [inline addons/web/static/src/views/calendar/calendar_common/calendar_common_renderer.js:249]: functions do not start with a capital letter

File: addons/web/static/src/views/calendar/calendar_renderer.scss [hunk 6]
```diff
@@ -0,0 +1,697 @@
+.o_calendar_view {
```
Comment (aab-odoo) [inline addons/web/static/src/views/calendar/calendar_renderer.scss:1]: you sure about this? this isn't the renderer node, so either the classname isn't correct, or these rules aren't in the correct file

File: addons/web/static/src/views/calendar/calendar_renderer_mobile.scss [hunk 7]
```diff
@@ -0,0 +1,34 @@
+@include media-breakpoint-down(md) {
+
+    .o_calendar_view {
```
Comment (aab-odoo) [inline addons/web/static/src/views/calendar/calendar_renderer_mobile.scss:3]: same

File: addons/web/static/src/views/calendar/date_picker/calendar_date_picker.js [hunk 8]
```diff
@@ -0,0 +1,115 @@
+/** @odoo-module **/
+
+//import { calculateWeekNumber } from "../date_utils";
+
+const { Component, onMounted, onWillUnmount, useEffect, useRef, xml } = owl;
+
+// This component uses JQuery!
+// Should we find another lib for date picker?
+// Or write our own date picker?
```
Comment (aab-odoo) [inline addons/web/static/src/views/calendar/date_picker/calendar_date_picker.js:9]: this is the plan, I think @Arcasias has a wip PR for that

File: addons/web/static/src/views/calendar/date_picker/calendar_date_picker.js [hunk 9]
```diff
@@ -0,0 +1,115 @@
+/** @odoo-module **/
+
+//import { calculateWeekNumber } from "../date_utils";
```
Comment (aab-odoo) [inline addons/web/static/src/views/calendar/date_picker/calendar_date_picker.js:3]: remove

File: addons/web/static/src/views/calendar/date_picker/calendar_date_picker.js [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+            this.$el.datepicker("destroy");
+            const picker = document.querySelector("#ui-datepicker-div:empty");
+            if (picker) {
+                picker.remove();
+            }
+        });
+    }
+
+    get dayNamesMin() {
+        // I think this is totally wrong!
```
Comment (aab-odoo) [inline addons/web/static/src/views/calendar/date_picker/calendar_date_picker.js:29]: lol

File: addons/web/static/src/views/calendar/date_picker/calendar_date_picker.js [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+    }
+    get options() {
+        return {
+            dayNamesMin: this.dayNamesMin,
+            firstDay: (this.props.model.firstDayOfWeek || 0) % 7,
+            monthNames: luxon.Info.months("short"),
+            onSelect: this.onDateSelected.bind(this),
+            showOtherMonths: true,
+            //calculateWeek: calculateWeekNumber, // JPP Je pense que c'est pas necesaire
+            // defaultDate: this.props.model.date.toFormat("yyyy-MM-dd"),
```
Comment (aab-odoo) [inline addons/web/static/src/views/calendar/date_picker/calendar_date_picker.js:43]: there are several comments in this file that should be considered

File: addons/web/static/src/views/calendar/date_picker/calendar_date_picker.js [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+        const model = this.props.model;
+        const date = luxon.DateTime.local(
+            +info.currentYear,
+            +info.currentMonth + 1,
+            +info.currentDay
+        );
+        let scale = "week";
+
+        if (model.date.hasSame(date, "day")) {
+            // const scales = model.scales.slice().reverse();
```
Comment (aab-odoo) [inline addons/web/static/src/views/calendar/date_picker/calendar_date_picker.js:93]: remove?

File: addons/web/static/src/views/calendar/date_utils.js [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+    deserializeDate,
+    deserializeDateTime,
+    serializeDate,
+    serializeDateTime,
+} from "@web/core/l10n/dates";
+
+// export function calculateWeekNumber(date) {
+//     //WOWL: USED ONLY FOR THE DATEPICKER (NEEDED ???)
+//     return luxon.DateTime.fromJSDate(date).weekNumber;
+// }
```
Comment (aab-odoo) [inline addons/web/static/src/views/calendar/date_utils.js:13]: clean

File: addons/web/static/src/views/calendar/date_utils.js [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+
+export function deserialize(value, type) {
+    return type === "date" ? deserializeDate(value) : deserializeDateTime(value);
+}
+export function serialize(value, type) {
+    return type === "date" ? serializeDate(value) : serializeDateTime(value);
+}
```
Comment (aab-odoo) [inline addons/web/static/src/views/calendar/date_utils.js:20]: not sure about this file. I would either get rid of those two helpers, or move them directly into dates

File: addons/web/static/src/views/fields/many2many_tags/many2many_tags_field.js [hunk 15]
```diff
@@ -354,3 +354,6 @@ Many2ManyTagsFieldColorEditable.extractProps = (params) => {
 };
 
 registry.category("fields").add("form.many2many_tags", Many2ManyTagsFieldColorEditable);
+
+registry.category("fields").add("calendar.one2many", Many2ManyTagsField);
+registry.category("fields").add("calendar.many2many", Many2ManyTagsField);
```
Comment (aab-odoo) [inline addons/web/static/src/views/fields/many2many_tags/many2many_tags_field.js:359]: is this necessary? I wasn't aware the calendar view looks in the field registry.

File: addons/web/static/tests/mobile/helpers.js [hunk 16]
```diff
@@ -0,0 +1,124 @@
+/** @odoo-module **/
+
+import { findElement, triggerEvent } from "../helpers/utils";
+
+async function swipe(target, selector, direction) {
```
Comment (aab-odoo) [inline addons/web/static/tests/mobile/helpers.js:5]: is this in the correct commit?

**Review by aab-odoo — COMMENTED** [review]
Excellent work. The PR already looks really strong (I didn't read the details, obviously)

File: addons/web/static/src/views/calendar/calendar_arch_parser.js (hunk 3 above)
Comment (jpp-odoo) [reply addons/web/static/src/views/calendar/calendar_arch_parser.js:182]: The `activeActions` creates edit, create, delete and duplicate searching on the attributes with the same names.
Here we only have two of these (create and delete), the edit is completely different
```js
canEdit = !fields[fieldMapping.date_start].readonly;
```
Do you think that we need to use `activeActions`?

**Comment by brboi** [comment]
robodoo method=rebase-ff

**Comment by Iucapad** [comment]
So cool, congrats 👏

**Review by aab-odoo — APPROVED** [review]
robodoo r+

good job @brboi @jpp-odoo and @mcm-odoo :+1:

**Pull request closed without merging**